### ISSUE TITLE
[MJAVADOC-123] Encoding Error When Getting project.name For JavaDoc -doctitle

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-javadoc-plugin#570]
**[Steven Coco](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=steevcoco)** opened **[MJAVADOC-123](https://issues.apache.org/jira/browse/MJAVADOC-123?redirect=false)** and commented

I have found an encoding handling error in JavaDoc generation.

The POM for this project is in UTF-8 encoding. the project.name element contains the trade mark character, U+2122.

When the JavaDoc plugin builds the documentation, the default HTML charset is used: ISO-8859-1.

When viewing the JavaDoc pages, the trade mark character in the title has been munged: "â¢".

At some location, the project name's text is being copied and ported to the JavaDoc tool in the wrong encoding.

The attachment has a test case. Since it's very small, the javadoc has already been built there to prevent any unwanted encoding mismatches during transmission of the attachment. View the POM to see the trade mark character in the name. View target/site/apidocs/index.html to see the munged trade mark character. View the src/main/java/overview.html file to see the other content used for the API overview page.

N.B.: In the attachment, all files are in UTF-8 encoding.

---

**Attachments:**
- [javadoc-encoding.tar.gz](https://issues.apache.org/jira/secure/attachment/12717079/javadoc-encoding.tar.gz) (_7.49 kB_)
- [javadoc-encoding-2.tar.gz](https://issues.apache.org/jira/secure/attachment/12717056/javadoc-encoding-2.tar.gz) (_7.50 kB_)
- [javadoc-encoding-3-4.tar.gz](https://issues.apache.org/jira/secure/attachment/12716910/javadoc-encoding-3-4.tar.gz) (_14.29 kB_)

**Comment by jira-importer** [comment]
**[Siveton Vincent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=siveton)** commented

You need to add in the \<configuration/> tag:

```
<charset>UTF-8</charset>
```

**Comment by jira-importer** [comment]
**[Steven Coco](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=steevcoco)** commented

Yikes!

It would appear that my test case was in error and this is not an issue. However, it is a bug.

In the first test case I used the trade mark character, U+2122. That character does not appear in the ISO-8859-1 charset, so you might expect something weird to happen. However if you look carefully you see that the HTML file contains a 3-byte sequence: the UTF-8 encoding of that character.

I changed the test case to make it more clear. Now I used the registered trade mark symbol, U+00AE. This character does appear in the ISO-8859-1 charset; as well as in the UTF-8 charset, where it is used. But still, the HTML file winds up containing a munged byte sequence: the raw 2-byte UTF-8 encoding for that character: 0xC2 0xAE.

It is clear that the plugin is not reading the POM in its declared encoding: likely, it's reading in ISO-8859-1 instead of UTF-8.

No matter what charsets are used in the POM or for the JavaDoc plugin, the files must be read in their declared encoding and transcoded during processing. So as in the first case where the character does not appear in the target charset, then any agreed-upon behavior could be acceptable: character replacement or perhaps something else. But if the character is mappable in the target charset then it must make it through.

I will try to take a look at the sources.

**Comment by jira-importer** [comment]
**[Steven Coco](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=steevcoco)** commented

javadoc-encoding-2

**Comment by jira-importer** [comment]
**[Siveton Vincent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=siveton)** commented

I tried your second attachment on Windows and on Solaris and all works here.

Some hits:
* try to use -Ddebug=true [1] to see that `@options` file is in UTF-8 and contains TM char
* specify to javadoc tool that the Java file encoding is UTF-8, i.e. -J-Dfile.encoding=UTF-8 [2]. In the Maven plugin, you need to use the additionalJOption parameter only available in the snapshot

Waiting for your feedback

[1] http://maven.apache.org/plugins/maven-javadoc-plugin/faq.html
[2] http://java.sun.com/j2se/1.4.2/docs/tooldocs/windows/javadoc.html#J

**Comment by jira-importer** [comment]
**[Steven Coco](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=steevcoco)** commented

This Web app is broken. I wrote a lengthy comment and apparently my session timed out. When I submitted I got an error message saying I was not alows to post. And I can't go "back" and get the text that I entered into the form. I have to start now from scratch and that's totally unacceptable.

[This version shorter because I'm out of time.]

I tried both options.

I made 2 new projects: version 3 and 4. They contain the plugin snapshot repository declaration.

Version 3 POM has been switched to ISO-8859-1: declared in the XML declaration, and correctly saved.

Building with this command:

mvn -U -X -e -Ddebug=true -Dfile.encoding=UTF-8 clean javadoc:javadoc

Verify that maven is using the JavaDoc snapshot plugin.

In each case the options file is in UTF-8 encoding. However, in the version 3 file, the (R) has been inserted as the Unicode replacement character (correctly encoded in UTF-8). This should never happen since that character exists in both charsets.

Tried switching to -Dfile.encoding=ISO-8859-1 and building both again -- both builds would still be valid builds this way.

Similar problem, only seemingly worse: options file is again in UTF-8 encoding (with the replacement character).

Something is using the wrong encoding.

I'm not sure when I'll get to look at the sources. I'll keep trying.

Ciao.

**Comment by jira-importer** [comment]
**[Steven Coco](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=steevcoco)** commented

JavaDoc encoding bug tests, versions 3 & 4.

**Comment by jira-importer** [comment]
**[Siveton Vincent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=siveton)** commented

> Building with this command:
> 
> mvn -U -X -e -Ddebug=true -Dfile.encoding=UTF-8 clean javadoc:javadoc

-Dfile.encoding=UTF-8 is specify to the Maven jvm **NOT** to the javadoc one. Specify this is similar to:

```
MAVEN_OPTS=-Dfile.encoding=UTF-8
JAVA_OPTS=-Dfile.encoding=UTF-8
```

You need to use **-J-Dfile.encoding=UTF-8** for the javadoc. To handle this, you need the snapshot from svn (not yet deployed on any snapshot repo) and add -J-Dfile.encoding=UTF-8 to the \<additionalJOption/> parameter.

HTH

**Comment by jira-importer** [comment]
**[Siveton Vincent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=siveton)** commented

Any news? Could we close it?

**Comment by jira-importer** [comment]
**[Steven Coco](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=steevcoco)** commented

Sorry, I've been so busy.

Since I see the problem when I use ISO-8859-1 in both the POM and the javadoc, I'd like to perform a test that does guarantee that there is no bug: I currently still think there's a problem since characters in 8859-1 should port around without being transcoded, but in one of the second examples the javadoc contains the replacement character; so I want to be sure about it.

I'll really try to test this. If I lag in time much longer I'll post back a comment.

Thanks.

**Comment by jira-importer** [comment]
**[Steven Coco](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=steevcoco)** commented

I'm sorry I let so much time pass on this!

I got it working: setting the JavaDoc plugin's encoding and charset to UTF-8 has fixed my problem.

I was assuming that since this platform's encoding is UTF-8 then Maven would be using that in some place, but I think I failed to see the real picture of what files are being created and the JavaDoc plugin's work flow. I think I was just confused: it's working this way. If you think this does seem right then that should clear this up.

Thanks for the help.

Ciao!
-Steev Coco.
